### ISSUE TITLE
Status updates occurring close together result in an exeception

### ISSUE THREAD

**Issue by shawkins** [java-operator-sdk/java-operator-sdk#409]
We have a custom resource with several operands.  The current code will return UpdateControl.updateStatusSubResource on every call to ResourceController.createOrUpdateResource.  When several operand updates are processed close together we sometimes see:

```
ERROR [io.jav.ope.pro.EventDispatcher] Error during event processing ExecutionScope
...
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PUT
...
the object has been modified; please apply your changes to the latest version and try again
```

I need to run in debug more to check if the operator sdk is retrying this - if it is retrying, then can this be logged at debug.  If it's not, then is this something that the operator or fabric8 should be handling?

**Comment by shawkins** [comment]
It does appear in debug that the retry handling is kicking in - the logic will look for any new events, and run the update again.  So this does appear to be benign.  @metacosm are you open to a change that would lower the amount of logging related to this exception?

**Comment by metacosm** [comment]
Yep.

**Comment by shawkins** [comment]
I've submitted a pr to change this to an info log without a stacktrace.

**Comment by metacosm** [comment]
Can we close this issue then? Or are there still things that need to be investigated?

**Comment by shawkins** [comment]
No, it seems that this is how it's expected to work.  The only alternative/addition would be to add locking per resource being passed to createOrUpate such that you could serialize it's processing - at least for the single operator case that would typically eliminate these conflicts.

**Comment by metacosm** [comment]
What do you think, @csviri?

**Comment by csviri-form3** [comment]
Hi @metacosm and @shawkins 
Not sure if I understand the situation correctly: We have already serialized processing in terms of operator for a custom resource. If I understand correctly what happens is that an other process(es) changes the CR frequently, that lead to an optimistick locking error. 

By locking I guess you mean pessimistic locking. I'm not aware that kubernetes supports such thing. I think what can be done however is not do updates on status subresource using optimistic locking, since the controller is anyways the only process that should change status.
See:
https://github.com/java-operator-sdk/java-operator-sdk/blob/b4c6483ab3c4ddffcdf030681d03c4f3258556f8/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/EventDispatcher.java#L203-L207

I think this needs to be changed in fabric8 client. @metacosm

**Comment by csviri** [comment]
Hi @metacosm and @shawkins
Not sure if I understand the situation correctly: We have already serialized processing in terms of operator for a custom resource. If I understand correctly what happens is that an other process(es) changes the CR frequently, that lead to an optimistick locking error.

By locking I guess you mean pessimistic locking. I'm not aware that kubernetes supports such thing. I think what can be done however is not do updates on status subresource using optimistic locking, since the controller is anyways the only process that should change status.
See:
https://github.com/java-operator-sdk/java-operator-sdk/blob/b4c6483ab3c4ddffcdf030681d03c4f3258556f8/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/EventDispatcher.java#L203-L207

I think this needs to be changed in fabric8 client. @metacosm

**Comment by csviri** [comment]
@shawkins just clarify 1 thing what do you mean by operand. Its a resource that is created by the controller and watched by an event source ?

Regarding to this comment:
https://github.com/java-operator-sdk/java-operator-sdk/blob/b4c6483ab3c4ddffcdf030681d03c4f3258556f8/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/DefaultEventHandler.java#L211

This is an edge case, with event sources where as described if we do an update and execute the controller  again because there were already events from different event source than custom resource event source, the cached custom resource could be out of the date (since we cache when it arrvies on a change from a watch). But this code should solved this problem in all cases. So if by actor you mean an event source, then yes this situation only apply when there are additional event sources.

**Comment by shawkins** [comment]
>  just clarify 1 thing what do you mean by operand. Its a resource that is created by the controller and watched by an event source ?

That is correct.  Our cr that the operator is managing creates another custom resource, configmaps, secrets, deployments, etc.  We watch those resources via an informer and register them as event sources.

For a given resource R we see logs similar to:

2021-04-24 21:43:12,309 DEBUG [org.bf2.ope.eve.ResourceEventSource] Update event received for io.fabric8.kubernetes.api.model.apps.Deployment x for resource R
...
2021-04-24 21:43:12,433 ERROR [io.jav.ope.pro.EventDispatcher] Error during event processing ExecutionScope "with different operand events for R that the previous"

**Comment by adam-sandor** [comment]
I checked in on this, and I agree with the opinion that you can't make the expectation that other processes won't be changing status. It's just how Kubernetes works. So optimistic locking with just an info log is the right solution imho.

**Comment by csviri** [comment]
It's not restricted in any ways, it's just how it ment that only the controller updates the status.  But did not see this anywhere expicitly, (maybe will do a round around this on K8S Slack just out of curiosity). But if you have some nice reference about this pls share @adam-sandor.

The thing is that optimistic locking it won't help basically in any scenarios regaring to this, since it is anyways overriden by the controller on the next run, and its always a result of a run. So while it makes sense to to optimictick locking on specs it does not make to status. (Note that status is a subresource, overrinding it does not affect other parts of resource)

But I think this is a very minor detail, it's more interested from the pholosphical aspect of k8s.

**Comment by shawkins** [comment]
As best as I can tell we are observing this with 

1. Custom Resource CR Revision 1 is in cache
2. ExecutionConsumer to update Custom Resource CR Revision 1 is queued
3. CR is updated to Revision 2
4. ExecutionConsumer runs and fails due to the conflict

This typically doesn't happen because there is little time between 2 and 4, but in some scenarios, such as rolling upgrades or mass deletions we see a flood of events such that some of the ExecutionConsumers have been queued with the stale cr.

It seems like a way to narrow this case even further would be to a cache get to refresh, or immediately reject, the ExecutionScope at the start of the run https://github.com/java-operator-sdk/java-operator-sdk/blob/112a8268155a9f6262e1238b592e9813b09e1557/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/ExecutionConsumer.java#L25

**Comment by csviri** [comment]
@shawkins , yes, this can naturally happen, its not necessary between 2 and 4. But actually can happen anytime the controller is is getting executed. And usually the controller execution is much longer then the time between 2 and 4 ( by 4 I mean its actually getting picked up by the executor thread.)

So if this is the case actually there is nothing to solve. This is expected. You are right that getting the latest custom resource from the cache at this point:
https://github.com/java-operator-sdk/java-operator-sdk/blob/112a8268155a9f6262e1238b592e9813b09e1557/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/ExecutionConsumer.java#L25

Would be little bit better, but also would make the event list and the actual custom resource in the scope inconsistent.
Basically this situation is one of the main reason is why we implemented automatic retries, since this can naturally happen.

**Comment by csviri** [comment]
And again the situation when we do retry, because of optimisit locking on status is more concerning. Because there we do retry, while we actually could avoid it and be still consistent. Typically in the situation when there is change on the CR but not on `.spec`.

**Comment by metacosm** [comment]
Do you think we should re-open the fabric8 issue to support PATCH?

**Comment by csviri** [comment]
@metacosm that issue was because of the finalizers mainly, since adding finalizer is always an "additive" operation we could do it by patching ("just add this finalizer to the finalizer list") request. But to be honest, I'm also quite happy with the current solution, maybe that would be a little bit nicer and straighforward. 

The patch also increased the resource version, so we have to think about that also after the request.  

I guess we can create an issue on this if you think it's worth to spend time on this part, but I'm fine with the current approach also. @metacosm

**Comment by csviri** [comment]
I will create a issue about the sub-resource update, I guess the k8s api does not require optimisitc locking there, (will verify). So we can disuss the pros and const regarding to that under the issue.

**Comment by shawkins** [comment]
>  I guess the k8s api does not require optimisitc locking there, (will verify)

This seems to require a resource version per the api server.  I've captured https://github.com/fabric8io/kubernetes-client/issues/3066 for fabric8 to consider this.
